---

_You are currently looking at **version 1.0** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [176]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [177]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [178]:
def graph_identification():
    %matplotlib notebook
    import matplotlib.pyplot as plt


#     G1 = P1_Graphs[0]
#     degrees1 = G1.degree()
#     degree_values1 = sorted(set(degrees1.values()))
#     hist1 = [list(degrees1.values()).count(i)/float(nx.number_of_nodes(G1)) for i in degree_values1]

#     G2 = P1_Graphs[1]
#     degrees2 = G2.degree()
#     degree_values2 = sorted(set(degrees2.values()))
#     hist2 = [list(degrees2.values()).count(i)/float(nx.number_of_nodes(G2)) for i in degree_values2]

#     G3 = P1_Graphs[2]
#     degrees3 = G3.degree()
#     degree_values3 = sorted(set(degrees3.values()))
#     hist3 = [list(degrees3.values()).count(i)/float(nx.number_of_nodes(G3)) for i in degree_values3]

#     G4 = P1_Graphs[3]
#     degrees4 = G4.degree()
#     degree_values4 = sorted(set(degrees4.values()))
#     hist4 = [list(degrees4.values()).count(i)/float(nx.number_of_nodes(G4)) for i in degree_values4]

#     G5 = P1_Graphs[4]
#     degrees5 = G5.degree()
#     degree_values5 = sorted(set(degrees5.values()))
#     hist5 = [list(degrees5.values()).count(i)/float(nx.number_of_nodes(G5)) for i in degree_values5]


#     plt.figure(figsize=(10,5))
#     plt.subplot(231)
#     plt.plot(degree_values1, hist1, 'o')
#     plt.xlabel('Degree')
#     plt.ylabel('Fraction of Nodes')
#     plt.xscale('log')
#     plt.yscale('log')
#     plt.show()

#     plt.subplot(232)
#     plt.plot(degree_values2, hist2, 'o')
#     plt.xlabel('Degree')
#     plt.ylabel('Fraction of Nodes')
#     plt.xscale('log')
#     plt.yscale('log')
#     plt.show()

#     plt.subplot(233)
#     plt.plot(degree_values3, hist3, 'o')
#     plt.xlabel('Degree')
#     plt.ylabel('Fraction of Nodes')
#     plt.xscale('log')
#     plt.yscale('log')
#     plt.show()

#     plt.subplot(234)
#     plt.plot(degree_values4, hist4, 'o')
#     plt.xlabel('Degree')
#     plt.ylabel('Fraction of Nodes')
#     plt.xscale('log')
#     plt.yscale('log')
#     plt.show()

#     plt.subplot(235)
#     plt.plot(degree_values5, hist5, 'o')
#     plt.xlabel('Degree')
#     plt.ylabel('Fraction of Nodes')
#     plt.xscale('log')
#     plt.yscale('log')
#     plt.show()
    
#     # Your Code Here
#     print('G1 ave. clustering is {}'.format(nx.average_clustering(G1)))
#     print('G1 ave. shortest paht length is {}'.format(nx.average_shortest_path_length(G1)))

#     print('G2 ave. clustering is {}'.format(nx.average_clustering(G2)))
#     print('G2 ave. shortest paht length is {}'.format(nx.average_shortest_path_length(G2)))

#     print('G3 ave. clustering is {}'.format(nx.average_clustering(G3)))
#     print('G3 ave. shortest paht length is {}'.format(nx.average_shortest_path_length(G3)))

#     print('G4 ave. clustering is {}'.format(nx.average_clustering(G4)))
#     print('G4 ave. shortest paht length is {}'.format(nx.average_shortest_path_length(G4)))

#     print('G5 ave. clustering is {}'.format(nx.average_clustering(G5)))
#     print('G5 ave. shortest paht length is {}'.format(nx.average_shortest_path_length(G5)))

    ans = ['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']
    return  ans # Your Answer Here

In [179]:
# graph_identification()

['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

In [181]:
# for g in P1_Graphs:
#     print(nx.info(g))
#     print('\n')

Name: barabasi_albert_graph(1000,2)
Type: Graph
Number of nodes: 1000
Number of edges: 1996
Average degree:   3.9920


Name: watts_strogatz_graph(1000,10,0.05)
Type: Graph
Number of nodes: 1000
Number of edges: 5000
Average degree:  10.0000


Name: watts_strogatz_graph(750,5,0.075)
Type: Graph
Number of nodes: 750
Number of edges: 1500
Average degree:   4.0000


Name: barabasi_albert_graph(750,4)
Type: Graph
Number of nodes: 750
Number of edges: 2984
Average degree:   7.9573


Name: watts_strogatz_graph(750,4,1)
Type: Graph
Number of nodes: 750
Number of edges: 1500
Average degree:   4.0000




---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagmentSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagmentSalary` indicates whether that person is receiving a managment position salary.

In [155]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a managment position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a managment salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a managment position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.75 or higher will receive full points.

Using your trained classifier, return a series of length 252 with the data being the probability of receiving managment salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [156]:
def salary_predictions():
    
    # Your Code Here
    df = pd.DataFrame(index=G.nodes())
    df['departement'] = pd.Series(nx.get_node_attributes(G, 'Department'))
    df['management'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))
    df['clustering'] = pd.Series(nx.clustering(G))
    df['degree'] = pd.Series(G.degree())
    df['cc'] = pd.Series(nx.closeness_centrality(G))
    df['bc'] = pd.Series(nx.betweenness_centrality(G, normalized=True, endpoints=False))
    df['pr'] = pd.Series(nx.pagerank(G, alpha=0.8))
    df['hub'] = pd.Series(nx.hits(G)[0])
    # df['autho'] = pd.Series(nx.hits(G)[1])
    # df.head()
    
    y = df['management']
    y_train = y[y.notnull()]

    X = df.drop(['management'], axis=1)
    X_train = X[df.management.notnull()]
    X_test = X[df.management.isnull()]

    from sklearn.ensemble import RandomForestClassifier 
    from sklearn.model_selection import cross_val_score

#     for depth in range(3,10):
#         rfc = RandomForestClassifier(n_estimators=100, max_depth=depth)
#         print(np.mean(cross_val_score(rfc, X_train, y_train, cv=5, scoring='roc_auc')))


    rfc = RandomForestClassifier(n_estimators=100, max_depth=4)
    rfc.fit(X_train, y_train)

    prob = rfc.predict_proba(X_test)
    ans = pd.Series([x[1] for x in prob], index = X_test.index)
    return ans # Your Answer Here

In [157]:
# salary_predictions()

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [158]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.75 or higher will receive full points.

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [159]:
def new_connections_predictions():
    
    # Your Code Here
    df = future_connections
    df['pa'] = [i[2] for i in nx.preferential_attachment(G, df.index)]
    df['cn'] = df.index.map(lambda person: len(list(nx.common_neighbors(G, person[0], person[1]))))
    df['jc'] = [i[2] for i in nx.jaccard_coefficient(G, df.index)]
    df['rai'] = [i[2] for i in nx.resource_allocation_index(G, df.index)]
    df['aai'] = [i[2] for i in nx.adamic_adar_index(G, df.index)]
    
    y = df['Future Connection']
    y_train = y[y.notnull()]

    X = df.drop(['Future Connection'], axis=1)
    X_train = X[y.notnull()]
    X_test = X[y.isnull()]
    
    from sklearn.ensemble import RandomForestClassifier 
    from sklearn.model_selection import cross_val_score

#     for depth in range(3,10):
#         rfc = RandomForestClassifier(n_estimators=10, max_depth=depth)
#         print(np.mean(cross_val_score(rfc, X_train, y_train, cv=3, scoring='roc_auc')))
    
    rfc = RandomForestClassifier(n_estimators=10, max_depth=5)
    rfc.fit(X_train, y_train)

    prob = rfc.predict_proba(X_test)
    ans = pd.Series([x[1] for x in prob], index = X_test.index)

    return ans # Your Answer Here

In [160]:
# new_connections_predictions()

(107, 348)    0.035166
(542, 751)    0.013140
(20, 426)     0.554884
(50, 989)     0.013140
(942, 986)    0.013140
(324, 857)    0.013140
(13, 710)     0.176434
(19, 271)     0.141148
(319, 878)    0.013140
(659, 707)    0.013140
(49, 843)     0.013140
(208, 893)    0.013140
(377, 469)    0.013185
(405, 999)    0.019584
(129, 740)    0.014809
(292, 618)    0.056315
(239, 689)    0.013140
(359, 373)    0.013185
(53, 523)     0.032244
(276, 984)    0.013140
(202, 997)    0.013140
(604, 619)    0.074984
(270, 911)    0.013140
(261, 481)    0.064172
(200, 450)    0.883555
(213, 634)    0.013140
(644, 735)    0.143287
(346, 553)    0.013140
(521, 738)    0.013140
(422, 953)    0.019270
                ...   
(672, 848)    0.013140
(28, 127)     0.974441
(202, 661)    0.013140
(54, 195)     0.995827
(295, 864)    0.013140
(814, 936)    0.013140
(839, 874)    0.013140
(139, 843)    0.013140
(461, 544)    0.013185
(68, 487)     0.013185
(622, 932)    0.013140
(504, 936)    0.016884
(479, 528) 